python/lsst/daf/butler/datastore/stored_file_info.py

```diff
@@ -320,3 +332,26 @@ def update(self, **kwargs: Any) -> StoredFileInfo:
 
     def __reduce__(self) -> str | tuple[Any, ...]:
         return (self.from_record, (self.to_record(),))
+
+
+class SerializedStoredFileInfo(_BaseModelCompat):
```

Comment:
StoredFileInfo isn't actually a dataclass, it's a subclass of the abstract base class StoredDatastoreItemInfo with a property that can't be instantiated without injecting configuration and several business logic methods.  And the representation of some of the fields varies between the various cases (e.g. component is `str | None` on the public API but in the database it's `str | Literal['__NULL_STRING__']`.)

I had considered making StoredFileInfo have-a SerializedStoredFileInfo and forwarding the properties to it, but it didn't seem to buy much because you still have to have the duplicate properties to do the forwarding.  I also think some of the representations might need to vary here (e.g. I kinda think SerializedStoredFileInfo shouldn't have a `path` at all, or if it does it should always be relative.  Whereas the one in the StoredFileInfo is sometimes absolute.)

I also considered making StoredFileInfo inherit-from SerializedStoreFileInfo or an additional shared pydantic model for the duplicated fields, but inheritance seems harder to understand than duplicating 5 string properties.  And if we add more properties, they will likely be non-nullable on StoredFileInfo but will have to be nullable on the serialized one to handle backwards compatibility.

I do think that the methods on StoredFileInfo could be moved to FileDatastore and StoredFileInfo could become an actual dataclass.  Or like if we were using the SqlAlchemy ORM layer StoredFileInfo could just be the ORM object.  But I know you guys had plans with exposing these more places which is why you added the abstract base class, so to my mind that kind of change is way outside the scope of this ticket.